src/event.rs

```diff
@@ -0,0 +1,101 @@
+//! Event subscriptions.
+
+use std::fmt;
+use std::time::Duration;
+
+/// A trait which is provided with information about events in a connection pool.
+pub trait HandleEvent: fmt::Debug + Sync + Send {
+    /// Called when a new connection is acquired.
+    ///
+    /// The default implementation does nothing.
+    #[allow(unused_variables)]
+    fn handle_acquire(&self, event: AcquireEvent) {}
+
+    /// Called when a connection is released.
+    ///
+    /// The default implementation does nothing.
+    #[allow(unused_variables)]
+    fn handle_release(&self, event: ReleaseEvent) {}
+
+    /// Called when a connection is checked out from the pool.
+    ///
+    /// The default implementation does nothing.
+    #[allow(unused_variables)]
+    fn handle_checkout(&self, event: CheckoutEvent) {}
+
+    /// Called when a checkout attempt times out.
+    ///
+    /// The default implementation does nothing.
+    #[allow(unused_variables)]
+    fn handle_timeout(&self, event: TimeoutEvent) {}
+
+    /// Called when a connection is checked back into the pool.
+    #[allow(unused_variables)]
+    fn handle_checkin(&self, event: CheckinEvent) {}
+}
+
+/// A `HandleEvent` implementation which does nothing.
+#[derive(Copy, Clone, Debug)]
+pub struct NopEventHandler;
+
+impl HandleEvent for NopEventHandler {}
+
+/// Information about an acquire event.
+#[derive(Debug)]
+pub struct AcquireEvent(pub(crate) ());
```

Comment:
I think what will be good form to add the comment about not used field when, it would seem it is possible to just use `struct AcquireEvent;`. Not all can understand at once that it is necessary in order that the event could not be created from the outside of a crate.